### PR TITLE
[P2P, Testing] Ensure node refuses unencrypted connections

### LINKED ISSUE #544: [P2P] Basic transport security
## Objective

Ensure transport-level encryption is enabled and required by default for all network participants.

The initial implementation of transport security should be compatible with the current formulation of V1 identity, this may be as simple as using the crypto helpers introduced in #534 together with [go-libp2p's `p2p/security/tls` package](https://pkg.go.dev/github.com/libp2p/go-libp2p@v0.26.0/p2p/security/tls). 

## Origin Document

image

## Goals

- P2P network peers support, and **require**, TLS with their counterparts.
- TLS configuration prefers a [GCM](https://en.wikipedia.org/wiki/Galois/Counter_Mode) variant of AES for symmetric encryption to optimize for speed and security.

## Deliverable

- [ ] ...
- [ ] A PR that enables configuring TLS security using hard-coded test fixtures
- [ ] Document the testing methodology in the source code so its repeatable (Bonus: automate)
- [ ] Unit tests for bad/unavailable certificates when turned on (Bonus: E2E tests)

## Non-goals / Non-deliverables

- Configuring remote infrastructure (i.e. DevNet/TestNet) to have certificates automatically available
- Changing any identity or crypto types and/or interfaces
- Manipulating X509 certificate(s) / chain(s)

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

1. Setup testing environment, following [development docs](https://github.com/pokt-network/pocket/tree/main/docs/development)
1. `docker exec -it node1.consensus ash` (see: [SO post](https://stackoverflow.com/questions/39362730/how-to-capture-packets-for-single-docker-container/61855965#61855965))
    1. (in container) `ip a`, note IP address
    2. (in container) `cat /sys/class/net/eth0/iflink`, note the "interface number"
    3. <Ctrl+d> or `exit` to exit the container shell
    4. (back on the host) `ip link|grep <INTERFACE_NUMBER>`, note the interface name, e.g. (if interface number was `55`, name would be `vetha165c5f@if54`):
        ```
        55: vetha165c5f@if54
        ```
1. Repeat 1.1-1.3 for `node2.consensus` container
1. Open wireshark (may need escalated privilege to see desired network interfaces)
1. Select the interface matching `node1.consensus`
1. Filter the captured packets by the IP matching `node2.consensus`
1. Look for a packet which has the `PSH` flag set
1. Inspect the `Data` contained in the package to see if any cleartext is readable (e.g. protobuf types: `...type.googleapis.com/consensus.HotstuffMessage...`)

- [ ] **Task specific tests or benchmarks**: `make ...`
- [ ] **New tests or benchmarks**: `make ...`
- [ ] **All tests**: `make test_all`
- [ ] **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @bryanchriswhite

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Description

Libp2p supports transport security via its [noise](https://noiseprotocol.org/) and [TLS](https://en.wikipedia.org/wiki/Transport_Layer_Security) and defaults to requiring peers to support at least one transport security protocol. This PR adds a test which asserts this behavior by starting up a P2P module and attempting to connect to it via a separate, insecure libp2p host.

## Issue

Fixes #544 

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [x] Other: Regression test

## List of changes

- Added a test which asserts that transport encryption is required (i.e. unencrypted connections are refused)

## Testing

- [x] `make develop_test`
- [ ] ~~[LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`~~

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [ ] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

**Review by Olshansk — CHANGES_REQUESTED** [review]
Just some NITs, but otherwise, great test!

File: p2p/transport_encryption_test.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	telemetryMock.EXPECT().SetBus(busMock).AnyTimes()
+	busMock.EXPECT().GetTelemetryModule().Return(telemetryMock).AnyTimes()
+
+	genesisStateMock := createMockGenesisState(keys[:1])
+	persistenceMock := preparePersistenceMock(t, busMock, genesisStateMock)
+	busMock.EXPECT().GetPersistenceModule().Return(persistenceMock).AnyTimes()
+
+	mockConsensusModule := mockModules.NewMockConsensusModule(ctrl)
+	mockConsensusModule.EXPECT().CurrentHeight().Return(uint64(1)).AnyTimes()
+	busMock.EXPECT().GetConsensusModule().Return(mockConsensusModule).AnyTimes()
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply p2p/transport_encryption_test.go:52]: I'm not sure it's quite that straightforward. I rearranged things to try to make it more readable and group things better.

**Review by dylanlott — APPROVED** [review]
Looks good to me! That was a sweet freebie to get from Libp2p.

**Review by Olshansk — CHANGES_REQUESTED** [review]
@bryanchriswhite Seems like this PR includes the python comment search code as well and is also based against the refactor.

Can we just rebase `p2p/transport_encryption_test.go` against main?

**Comment by bryanchriswhite** [comment]
> @bryanchriswhite Seems like this PR includes the python comment search code as well and is also based against the refactor.
> 
> Can we just rebase `p2p/transport_encryption_test.go` against main?

:thinking: Oops. Sorry, I'm not sure how that happened. :facepalm: 

On `main`, I checked out the test (`git checkout feat/basic-tls -- p2p/transport_encryption_test.go`) with the intention of trying to run it but I wasn't able to get past `make develop_start`:
```
go: finding module for package github.com/pokt-network/pocket/p2p/protocol
go: finding module for package github.com/pokt-network/pocket/p2p/utils
github.com/pokt-network/pocket/p2p tested by
        github.com/pokt-network/pocket/p2p.test imports
        github.com/pokt-network/pocket/p2p/protocol: no matching versions for query "latest"
github.com/pokt-network/pocket/p2p tested by
        github.com/pokt-network/pocket/p2p.test imports
        github.com/pokt-network/pocket/p2p/utils: no matching versions for query "latest"
make[1]: *** [Makefile:108: go_clean_deps] Error 1
make[1]: Leaving directory '/home/bwhite/Projects/pokt/pocket'
make: *** [Makefile:129: develop_start] Error 2
```

This change is indeed based off of #576. To resolve my mistake, I've dropped the offending commits and re-merged with the upstream branch.

**Review by Olshansk — APPROVED** [review]

**Pull request merged**